clang/lib/Sema/Metafunctions.cpp

```diff
@@ -2772,7 +2777,12 @@ bool is_variable(APValue &Result, Sema &S, EvalFn Evaluator, QualType ResultTy,
 
   bool result = false;
   if (R.getReflection().getKind() == ReflectionValue::RK_declaration) {
-    result = isa<const VarDecl>(R.getReflectedDecl());
+    Decl *D = R.getReflectedDecl();
+    if (auto *BD = dyn_cast<BindingDecl>(D)) {
+      result = (BD->getHoldingVar() != nullptr);
```

Comment:
Decided to go this way because of documentation:
```cpp
  /// Get the variable (if any) that holds the value of evaluating the binding.
  /// Only present for user-defined bindings for tuple-like types.
  VarDecl *getHoldingVar() const;
```

Reply:
Upon more careful reading of the Standard, I think I led you astray - we should not return `true` for `is_variable` in any of these cases.

I realized this while attempting to verify that `getHoldingVar` was the right API for this task, during which I came upon [this code](https://github.com/bloomberg/clang-p2996/blob/bbb6343672fbaabd2236b634fb0d382533eb7f19/clang/lib/Sema/SemaDeclCXX.cpp#L1333-L1362) which implements tuple-like binding decomposition in clang. The synthesized variable `RefVD` corresponds to the variable
```cpp
    S U_i r_i = initializer ;
```
mandated by [`[dcl.struct.bind]/4`](http://www.eel.is/c++draft/dcl.struct.bind#4), but notice that it's never stored as a property of the `BindingDecl B` that's being constructed. Instead, the `Sema::BuildDeclarationNameExpr()` function is used to synthesize an expression _referencing_ the variable, and that expression is set as the "binding" via `B->setBinding(E)`.

Looking back at the Standard, this [lines up](http://www.eel.is/c++draft/dcl.struct.bind#4.sentence-9):
> Each `v_i` is **the name of an lvalue of type `T_i` that refers to the object bound to `r_i`**; the referenced type is `T_i`.

So even though a variable _is_ synthesized for each binding in a tuple-like decomposition, I believe the binding itself is still _not_ a variable, but rather a named expression referencing the synthesized variable.

Sorry for the mixup!

Reply:
Thanks for deep dive! I reverted changes to implementation but left existing tests